pkg/cmds/restore_volumesnapshot.go

```diff
@@ -0,0 +1,191 @@
+			config, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfigPath)
+			if err != nil {
+				log.Fatalf("Could not get Kubernetes config: %s", err)
+			}
+			opt.kubeClient = kubernetes.NewForConfigOrDie(config)
+			opt.stashClient = cs.NewForConfigOrDie(config)
+			opt.snapshotClient = vs_cs.NewForConfigOrDie(config)
+
+			err = opt.RestoreVolumeSnapshot()
+			if err != nil {
+				log.Fatal(err)
+			}
+		},
+	}
+	cmd.Flags().StringVar(&masterURL, "master", "", "The address of the Kubernetes API server (overrides any value in kubeconfig)")
+	cmd.Flags().StringVar(&kubeconfigPath, "kubeconfig", "", "Path to kubeconfig file with authorization information (the master location is set by the master flag).")
+	cmd.Flags().StringVar(&opt.name, "restoresession.name", "", "Set Restore Session Name")
+	return cmd
+}
+
+func (opt *VSoption) RestoreVolumeSnapshot() error {
+	// Start clock to measure total session duration
+	startTime := time.Now()
+
+	restoreSession, err := opt.stashClient.StashV1beta1().RestoreSessions(opt.namespace).Get(opt.name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if restoreSession == nil || restoreSession.Spec.Target == nil {
+		return fmt.Errorf("restoreSession or  restoreSession target is nil")
+	}
+	fmt.Println("target", restoreSession.Spec.Target)
+
+	if restoreSession.Spec.Target.Replicas == nil {
+		pvcData = []PVC{}
+		for _, vol := range restoreSession.Spec.Target.VolumeClaimTemplates {
+			pvcData = append(pvcData, PVC{podOrdinal: types.Int32P(0), pvcName: vol.Name, pvc: &vol})
+
+		}
+	} else {
+		pvcData = []PVC{}
+		for i := int32(0); i < types.Int32(restoreSession.Spec.Target.Replicas); i++ {
+			for _, vol := range restoreSession.Spec.Target.VolumeClaimTemplates {
+				pvcData = append(pvcData, PVC{pvcName: vol.Name, podOrdinal: types.Int32P(i), pvc: &vol})
+			}
+		}
+
+	}
+
+	objectMeta := []metav1.ObjectMeta{}
+
+	for _, data := range pvcData {
+		pvc := opt.getPVCDefinition(data)
+		_, err := opt.kubeClient.CoreV1().PersistentVolumeClaims(opt.namespace).Get(data.pvc.Name, metav1.GetOptions{})
+		if err != nil {
+			pvc, err := opt.kubeClient.CoreV1().PersistentVolumeClaims(opt.namespace).Create(pvc)
+			if err != nil {
+				return err
+			}
+			objectMeta = append(objectMeta, pvc.ObjectMeta)
+
+		} else {
+			// write failure event
+			return opt.setRestoreSessionFailed(restoreSession, fmt.Sprintf("PVC %s is already exists", data.pvc.Name))
+		}
+	}
+
+	for i, pvc := range pvcData {
+		storageClass, err := opt.kubeClient.StorageV1().StorageClasses().Get(types.String(pvc.pvc.Spec.StorageClassName), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if *storageClass.VolumeBindingMode != storage_api_v1.VolumeBindingImmediate {
+			// write failure event
+			return opt.setRestoreSessionFailed(restoreSession, fmt.Sprintf("VolumeBindingMode is equal to %s because of phase is Pending", storage_api_v1.VolumeBindingWaitForFirstConsumer))
```

Comment:
Don't fail `RestoreSession` directly. Just update status for this host with appropriate error message. Set phase to `Unknown` instead of `Failed`.

Reply:
there is no constant for HostRestorePhase Unknown, so phase should be Failed